keps/973-workload-priority/README.md

```diff
@@ -0,0 +1,331 @@
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories
+
+Kueue issue [973](https://github.com/kubernetes-sigs/kueue/issues/973) provides details on the initial feature implementation.
+
+#### Story 1
+
+In an organization, admins want to set a lower priority for development workloads and a higher priority for production workloads.
+In such cases, they create two `WorkloadPriorityClass` and apply each one to the respective workloads.
+
+#### Story 2
+
+An organization desires to modify the priority of workloads that remain inactive for a specific duration.
+By developing a custom controller to manage Priority value of `WorkloadPriorityClass`, this expectation can be met.
+
+### Risks and Mitigations
+
+It's possible that the pod's priority conflicts with the workload's priority.
+For example, a high-priority job with low-priority pods may never run to completion because it may always be preempted by kube-scheduler.
+We should document the risks of pod preemption to uses.
+We can also point users to create `PriorityClass` for their pods that are [non-preempting](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#non-preempting-priority-class).
+If a workload's priority is high and pod's priority is low and the kube-scheduler initiates preemption, the pod's priority is prioritized. To prevent this behavior, non-preempting setting is needed.
+
+
+## Design Details
+
+### Kueue WorkloadPriorityClass API
+
+We introduce the `WorkloadPriorityClass` API.
+
+```golang
+type WorkloadPriorityClass struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Value int32   `json:"value,omitempty"`
+	Description string `json:"description,omitempty"`
+}
+
+```
+
+### How to use WorkloadPriorityClass on Job
+
+The `workloadPriorityClass` is specified through a label `kueue.x-k8s.io/workload-priority-class`.
+
+```yaml
+# sample-priority-class.yaml
+apiVersion: kueue.x-k8s.io/v1alpha1
+kind: WorkloadPriorityClass
+metadata:
+  name: sample-priority
+value: 10000
+description: "Sample priority"
+---
+# sample-job.yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  generateName: sample-job-
+  labels:
+    kueue.x-k8s.io/queue-name: user-queue
+    kueue.x-k8s.io/workload-priority-class: sample-priority
+spec:
+  parallelism: 3
+  completions: 3
+  suspend: true
+  template:
+    spec:
+      containers:
+      - name: dummy-job
+        image: gcr.io/k8s-staging-perf-tests/sleep:latest
+      restartPolicy: Never
+```
+
+The following workload is generated by the yaml above.
+The `PriorityClass` field can accept both k8s `PriorityClass` and `workloadPriorityClass` as values.
+To distinguish, when using `workloadPriorityClass`, a prefix "kueue.x-k8s.io/workload-priority-class/" is added.
```

Comment:
I am not sure this design is good

Reply:
I think it's better to have a separate field indicating where is the priority coming from:

`priorityClassSource: kueue.x-k8s.io/WorkloadPriorityClass`
or
`priorityClassSource: scheduling.k8s.io/PriorityClass`

Reply:
Updated.